media/2021-05-14-paying-homage-to-the-compall-michael-pattern/5-scala3-opaque-types.worksheet.sc

```diff
@@ -0,0 +1,47 @@
+// Approach using Scala 3 Opaque Type Aliases
+
+object Units {
+  opaque type Kilometers = Double
+  opaque type Miles = Double
+
+  val ZeroKm: Kilometers = 0
+  val ZeroMi: Miles = 0
+
+  def kilometres(value: Double): Option[Kilometers] = if (value < 0) None else Some(value)
+  def miles(value: Double): Option[Miles] = if (value < 0) None else Some(value)
+
+  def add(km1: Kilometers, km2: Kilometers): Kilometers = km1 + km2
+  def toKilometers(miles: Miles): Kilometers = miles * 1.6
+
+  extension (km: Kilometers) {
+    def + (km2: Kilometers): Kilometers = add(km, km2)
+  }
+
+  extension (miles: Miles) {
+    def toKm: Kilometers = toKilometers(miles)
+  }
+}
+
+import Units._
+
+class Booster() {
+  def provideLaunchBoost(): Miles = miles(100).getOrElse(ZeroMi)
+}
+
+class Rocket(booster: Booster) {
+  private var distance: Kilometers = ZeroKm
+
+  def launch(): Unit = {
+    // Kilometers and Miles are different types. So compiler prevents the previous bug
+    distance += booster.provideLaunchBoost().toKm
+  }
+
+  def distanceTravelled: Kilometers = distance
+}
+
+// For fun, let's make use of Scala 3's Univeral Apply Methods to omit the 'new'
+val rocket: Rocket = Rocket(Booster())
```

Comment:
Sort of a nitpick, but I think we loose out on the opportunity to show really show of Universal Apply Methods here since `Booster` could have been a case class this whole time, which would have also given you an apply.

```suggestion
// For fun, let's make use of Scala 3's Universal Apply Methods to omit the 'new'
val rocket: Rocket = Rocket(Booster())
```